lib/phoenix/verified_routes.ex

```diff
@@ -0,0 +1,757 @@
+  defp verify_query(_other, route, _acc) do
+    raise_invalid_query(route)
+  end
+
+  defp raise_invalid_query(route) do
+    raise ArgumentError,
+          "expected query string param to be compile-time map or keyword list, got: #{Macro.to_string(route)}"
+  end
+
+  @doc """
+  Generates an integrity hash to a static asset given its file path.
+  """
+  def static_integrity(%Plug.Conn{private: %{phoenix_endpoint: endpoint}}, path) do
+    static_integrity(endpoint, path)
+  end
+
+  def static_integrity(%_{endpoint: endpoint}, path) do
+    static_integrity(endpoint, path)
+  end
+
+  def static_integrity(endpoint, path) when is_atom(endpoint) do
+    endpoint.static_integrity(path)
+  end
+
+  @doc false
+  def __encode_query__(dict) when is_list(dict) or is_map(dict) do
+    case Plug.Conn.Query.encode(dict, &to_param/1) do
+      "" -> ""
+      query_str -> query_str
+    end
+  end
+
+  def __encode_query__(val), do: val |> to_param() |> URI.encode_www_form()
+
+  defp to_param(int) when is_integer(int), do: Integer.to_string(int)
+  defp to_param(bin) when is_binary(bin), do: bin
+  defp to_param(false), do: "false"
+  defp to_param(true), do: "true"
+  defp to_param(data), do: Phoenix.Param.to_param(data)
+
+  defp route_type(router, statics, test_path) do
+    cond do
+      static_path?(test_path, statics) -> :static
+      match_route?(router, test_path) -> :match
+      true -> :error
+    end
+  end
+
+  defp match_route?(router, test_path) when is_binary(test_path) do
+    split_path = for segment <- String.split(test_path, "/"), segment != "", do: segment
+    match_route?(router, split_path)
+  end
+
+  defp match_route?(router, split_path) when is_list(split_path) do
+    case router.__match_route__(split_path) do
+      {_forward_plug, false = _warn_on_verify?, _} -> false
+      {nil = _forward_plug, true = _warn?, _} -> true
+      {forward_plug, true = _warn?, _} -> match_forward_route?(router, forward_plug, split_path)
+      :error -> false
+    end
+  end
+
+  defp match_forward_route?(router, forward_router, split_path) do
+    if function_exported?(forward_router, :__routes__, 0) do
+      script_name = router.__forward__(forward_router)
+      match_route?(forward_router, split_path -- script_name)
+    else
+      true
+    end
+  end
+
+  defp build_route(route_ast, og_ast, env, endpoint_ctx, router) do
+    statics = Module.get_attribute(env.module, :phoenix_verified_statics, [])
+
+    router =
+      case Macro.expand(router, env) do
+        mod when is_atom(mod) ->
+          mod
+
+        other ->
+          raise ArgumentError, """
+          expected router to be to module, got: #{inspect(other)}
+
+          If you want to generate a compile-time router, use unverified_path/2 instead.
+          """
+      end
+
+    statics =
+      case Macro.expand(statics, env) do
+        list when is_list(list) -> list
+        other -> raise ArgumentError, "expected statics to be a list, got: #{inspect(other)}"
```

Comment:
Statics should already be expanded at this point, pls move the validation to `__using__/2`.